timemachine/fe/single_topology.py

```diff
@@ -1393,6 +1130,184 @@ def _align_nonbonded_pair_list(self):
+            maxes=None,
+            cutoff=src_nb_ixn_group.potential.cutoff,
+            beta=src_nb_ixn_group.potential.beta,
+        )
+
+    def _align_nonbonded_all_pairs(
+        self,
+        src_nb_all_pairs: BoundPotential[Nonbonded],
+        dst_nb_all_pairs: BoundPotential[Nonbonded],
+        combined_atom_map_mixin: AtomMapMixin,
+    ):
+        assert dst_nb_all_pairs.potential.beta == dst_nb_all_pairs.potential.beta
+        assert dst_nb_all_pairs.potential.cutoff == dst_nb_all_pairs.potential.cutoff
+
+        assert src_nb_all_pairs.potential.atom_idxs is not None
+        assert dst_nb_all_pairs.potential.atom_idxs is not None
+
+        src_atom_idxs = src_nb_all_pairs.potential.atom_idxs
+        dst_atom_idxs = dst_nb_all_pairs.potential.atom_idxs
+
+        atom_idxs, src_params, dst_params = self._process_ixn_group_idxs_and_params(
+            src_atom_idxs,
+            src_nb_all_pairs.params,
+            dst_atom_idxs,
+            dst_nb_all_pairs.params,
+            combined_atom_map_mixin.a_to_c,
+            combined_atom_map_mixin.b_to_c,
+            src_nb_all_pairs.potential.cutoff,
+        )
+
+        aligned_src_params = jnp.zeros((combined_atom_map_mixin.get_num_atoms(), 4))
+        aligned_src_params = aligned_src_params.at[atom_idxs].set(src_params)
+        aligned_dst_params = jnp.zeros((combined_atom_map_mixin.get_num_atoms(), 4))
+        aligned_dst_params = aligned_dst_params.at[atom_idxs].set(dst_params)
+
+        # assert that exclusion idxs and scale factors are identical at the end-states. We have no way
+        # of morphing scale factors since they're not part of the "parameters", and HREX requires that we use
+        # .set_params to do all interpolation schemes.
+
+        src_exclusion_idxs = src_nb_all_pairs.potential.exclusion_idxs
+        dst_exclusion_idxs = dst_nb_all_pairs.potential.exclusion_idxs
+
+        src_exclusion_scale_factors = src_nb_all_pairs.potential.scale_factors
+        dst_exclusion_scale_factors = dst_nb_all_pairs.potential.scale_factors
+
+        exclusions_kv = dict()
+
+        for (i, j), sf in zip(src_exclusion_idxs, src_exclusion_scale_factors):
+            # mypy complaint: error: Value of type variable "SupportsRichComparisonT" of "sorted" cannot be "ndarray[tuple[int, ...], dtype[signedinteger[_32Bit]]]"  [type-var]
+            key = tuple(sorted([combined_atom_map_mixin.a_to_c[i], combined_atom_map_mixin.a_to_c[j]]))  # type: ignore
+            exclusions_kv[key] = sf
+
+        for (i, j), sf in zip(dst_exclusion_idxs, dst_exclusion_scale_factors):
+            # mypy complaint: error: Value of type variable "SupportsRichComparisonT" of "sorted" cannot be "ndarray[tuple[int, ...], dtype[Any]]"  [type-var]
+            key = tuple(sorted([combined_atom_map_mixin.b_to_c[i], combined_atom_map_mixin.b_to_c[j]]))  # type: ignore
+            if key in exclusions_kv:
+                np.testing.assert_array_equal(exclusions_kv[key], sf)
```

Comment:
This is a hard requirement - we require that any duplicate exclusions in `src` and `dst` share the same scaling factors (since we're not allowed interpolate over them currently with our HREX `.set_params()` requirement).

Reply:
Can you add this as a comment in the code?

Reply:
addressed in https://github.com/proteneer/timemachine/pull/1538/commits/1f5f30c9e58a747173bba292635d8b4e040dca91

Reply:
> we require that any duplicate exclusions in src and dst share the same scaling factors

question/unsure: is this needed given that we already have a check ensuring that the (unbound) potentials for all states are identical?

https://github.com/proteneer/timemachine/blob/71a20bd61400f72d12ae09ace74b9c1939667056/timemachine/fe/free_energy.py#L1430-L1433

Reply:
> I don't think it hurts to be defensive and try and catch this kind of behavior a little bit early

Agreed; mostly just wanted to check my understanding